design.md

```diff
@@ -0,0 +1,138 @@
+  - List files in a repo.
+  - Read the contents of a file in a repo.
+  - Run a script in a repo.
+    - May pass arguments as a list of objects.
+    - May receive a return value as an object.
+  - Load a robot from a robot config file in a repo.
+
+#### Hardware Management
+
+- Operations:
+  - Load a robot from its config file.
+    - May specify the contents of the config file.
+    - May specify a Git File URI.
+    - Must specify whether to connect to hardware or a simulator.
+      - In the hardware case, a connection method must be specified.
+      - In the simulator case, a simulator plugin must be specified.
+
+#### Simulator Plugins
+
+- Operations:
+  - List available plugins
+  - Add a plugin
+    - Adding a plugin with an ID that conflicts with an ID of an already-added plugin throws an error.
+    - ! RFC: What should the schema for adding a plugin be? The most natural method is by specifying the path to the Jar file, but we don't want to use file paths. We could support specifying a Git File URI. What else? I don't want to allow specifying content because that would involve sending potentially lots of megabytes over the RPC call. We could have a registry of simulator plugins similar to how we will have a registry of toolchains.
+  - Remove a plugin by its ID.
+- Each plugin has a unique ID and a Jar file.
+- The kernel must be able to download simulator plugins.
+  - We can ship the default simulator plugin with the IntelliJ plugin, but there must be a way to download any of them because of the [usability requirement](#usability).
+
+#### UI Interaction
+
+- ! RFC: The kernel must be able to notify the UI of multiple tasks happening in parallel. For example, running a script could cause some repos to get cloned and a toolchain to get downloaded in parallel. As a user, I want to see both of these operations as tasks in IntelliJ happening in parallel. As a general rule, a script can start multiple tasks.
+- ! RFC: The kernel needs a way to ask the UI for confirmation with a custom confirmation message. For example, if the kernel needs to flash the device with a new program, by default it should ask the user for confirmation. There also needs to be some local environment configuration option that bypasses this so that the confirmation results is either always allowed or denied (for headless operation). For example, you may want to always allow the kernel to flash the device during development or during some competitions (as a user, I would rather have a delayed start to my match rather than having the robot not work at all; it depends on the competition). As another example, you may want to always stop the kernel from flashing the device if you have released a product and you expect all the products to work the same for all your users and would rather handle these issues with your own support system.
+
+#### Timeout
+
+If no script is running and no gRPC interaction has occurred for some time period, then the kernel should exit.
```

Comment:
Is there any possibility of loss of user data in this case? What about a remote kernel; how will it be started?

Also, why? Is this just a failsafe in case the client hard exits and leaves it hanging? In that case, should the client be sending keepalives over gRPC (good practice anyway)?

Reply:
> Is there any possibility of loss of user data in this case?

Yes.

> Also, why? Is this just a failsafe in case the client hard exits and leaves it hanging?

That is one case. More generally, we are writing a daemon, so there is a chance that it could be "leaked" and left running erroneously. Therefore, the kernel needs to be able to close itself if it isn't being "used".

I have rewritten this to use a keepalive call.